providers/apis/defi/osmosis/price_fetcher.go

```diff
@@ -0,0 +1,198 @@
+	if logger == nil {
+		return nil, fmt.Errorf("logger cannot be nil")
+	}
+
+	if err := api.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("invalid api config: %w", err)
+	}
+
+	if api.Name != Name {
+		return nil, fmt.Errorf("invalid api name; expected %s, got %s", Name, api.Name)
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("api is not enabled")
+	}
+
+	if apiMetrics == nil {
+		return nil, fmt.Errorf("metrics cannot be nil")
+	}
+
+	return &APIPriceFetcher{
+		api:               api,
+		client:            client,
+		logger:            logger.With(zap.String("fetcher", Name)),
+		metaDataPerTicker: newMetadataCache(),
+	}, nil
+}
+
+// Fetch fetches prices from the osmosis API for the given currency-pairs. Specifically
+// for each currency-pair,
+//   - Query the spot price.
+func (pf *APIPriceFetcher) Fetch(
+	ctx context.Context,
+	tickers []oracletypes.ProviderTicker,
+) oracletypes.PriceResponse {
+	resolved := make(oracletypes.ResolvedPrices)
+	unresolved := make(oracletypes.UnResolvedPrices)
+
+	wg := sync.WaitGroup{}
+	unresolvedMtx := sync.Mutex{}
+	resolveMtx := sync.Mutex{}
+	wg.Add(len(tickers))
+
+	pf.logger.Info("fetching for tickers", zap.Any("tickers", tickers))
+
+	// make sure metadata cache is set properly
+	for _, ticker := range tickers {
+		_, found := pf.metaDataPerTicker.getMetadataPerTicker(ticker)
+		if !found {
+			_, err := pf.metaDataPerTicker.updateMetaDataCache(ticker)
+			if err != nil {
+				pf.logger.Debug("failed to update metadata cache", zap.Error(err))
+			}
+		}
+	}
+
+	for _, ticker := range tickers {
+		go func(ticker oracletypes.ProviderTicker) {
+			var err error
+			defer wg.Done()
+
+			metadata, found := pf.metaDataPerTicker.getMetadataPerTicker(ticker)
+			if !found {
+				unresolvedMtx.Lock()
+				defer unresolvedMtx.Unlock()
+				unresolved[ticker] = providertypes.UnresolvedResult{
+					ErrorWithCode: providertypes.NewErrorWithCode(
+						NoOsmosisMetadataForTickerError(ticker.String()),
+						providertypes.ErrorTickerMetadataNotFound,
+					),
+				}
+				return
+			}
+
+			callCtx, cancel := context.WithTimeout(ctx, pf.api.Timeout)
+			defer cancel()
+
+			resp, err := pf.client.SpotPrice(callCtx,
+				metadata.PoolID,
+				metadata.BaseTokenDenom,
+				metadata.QuoteTokenDenom,
+			)
+			if err != nil {
+				unresolvedMtx.Lock()
+				defer unresolvedMtx.Unlock()
+				pf.logger.Error("failed to fetch spot price", zap.Error(err))
+				unresolved[ticker] = providertypes.UnresolvedResult{
+					ErrorWithCode: providertypes.NewErrorWithCode(
+						err,
+						providertypes.ErrorAPIGeneral,
+					),
+				}
+				return
+			}
+
+			price, err := calculatePrice(resp)
+			if err != nil {
+				pf.logger.Error("failed parse spot price response", zap.Error(err))
+				unresolved[ticker] = providertypes.UnresolvedResult{
+					ErrorWithCode: providertypes.NewErrorWithCode(
+						err,
+						providertypes.ErrorFailedToParsePrice,
+					),
+				}
+				return
+			}
```

Comment:
shouldn't this have the unresolvedMtx lock?

maybe we should replace this logic with a closure?

```go
writeToUnresolved := func(ticker, err, code) {
    ....
}
```

Reply:
yo @aljo242 any thoughts on the closure idea above?

Reply:
pushed here https://github.com/skip-mev/slinky/pull/607/commits/bb7b13dd2d1d9a118c645a18ae70d789dbc6e577